anvil/src/shell/x11.rs

```diff
@@ -44,9 +45,126 @@ impl OldGeometry {
     }
 }
 
+impl<BackendData: Backend> XWaylandShellHandler for CalloopData<BackendData> {
+    fn xwayland_shell_state(&mut self) -> &mut XWaylandShellState {
+        &mut self.state.xwayland_shell_state
+    }
+}
+
+impl<BackendData: Backend> XWaylandShellHandler for AnvilState<BackendData> {
```

Comment:
I don't think we want a second trait and force downstream to awkwardly forward calls like this. I think it is fine to just drop the old matching logic and require a recent enough Xwayland version supporting this protocol.

Reply:
This hack predates my PR, I just made it explicit. I'll try to explain. It has nothing to do with the old matching logic.

This is necessary because Anvil has an `EventLoop<CalloopData>` but a different state object (`AnvilState`), on which most handlers are implemented. Because `Xwm` requires `LoopHandle<D> where D: XwmHandler`, the existing code implemented `XwmHandler` directly on `CalloopData` instead of `AnvilState`. But then it had to call into the Xwm from one of those other handlers on `AnvilState`:

https://github.com/Smithay/smithay/blob/418190e4992ce642e6bac873307d4fc4fa9a1e89/anvil/src/shell/mod.rs#L142

I changed the commit hook to take a `D` as a parameter, so that broke. I actually think this hack is cleaner, in a way, although it's terrible on its face.

Note that this all goes away if remove the tight integration with calloop that xwm has. I would propose doing that by adding a `poll_fd()` and a `dispatch` method, mirroring the wayland side.

By the second trait, do you mean `XWaylandShellHandler`? I think having that separated makes sense, since it's just for the protocol itself. And that means people can supply their own xwm implementation.

Reply:
> Note that this all goes away if remove the tight integration with calloop that xwm has. I would propose doing that by adding a poll_fd() and a dispatch method, mirroring the wayland side.

Good idea.

> By the second trait, do you mean XWaylandShellHandler? I think having that separated makes sense, since it's just for the protocol itself. And that means people can supply their own xwm implementation.

Then maybe we need to split it. But the duplicated x-wm callbacks are not nice to use.

Reply:
Sorry, I'm a bit lost. What do you mean by duplicated callbacks?

Reply:
Having the `XWaylandShellHandler` call into `XwmHandler`.

Reply:
> Having the XWaylandShellHandler call into XwmHandler.

I don't see where it does that. The two traits are independent, except that `XwmHandler` requires `D` to also implement `XWaylandShellHandler`, since `X11wm` uses both to do the matching. (I should actually move the constraint to `X11Wm` to make that clearer, on reflection, but that's a trivial change).

What's happening here (where this comment is) is that we're implementing `XWaylandShellHandler` for both `CalloopData<BackendData>` and `AnvilState<BackendData`. The forwarding is just between two implementations of the same trait, because of an interaction between `X11Wm` weirdness and `Anvil` weirdness. There's no cross-pollination between `XWaylandShellHandler` and `XwmHandler`.

I'm really sorry if I'm just being thick! I have a head cold at the moment and having trouble focusing :)

Reply:
> I don't see where it does that.

Ah my god, the diff view had me confused... sorry!

Yeah it was the two `XwmHandler` implementations (once for `CalloopData<AnvilState>` and the other for `AnvilState`), that were throwing me off.

I don't think we can really keep those in anvil, even if a normal compositor wouldn't have to do this, because this encourages people to just copy this hacky approach.

We should get rid of CalloopData in anvil, is that something you feel like possibly tackling?

Reply:
> I should actually move the constraint to X11Wm to make that clearer, on reflection, but that's a trivial change

Addressed in 6c675b5.

> We should get rid of CalloopData in anvil, is that something you feel like possibly tackling?

Between that and removing `X11Wm`'s dependency on `LoopHandle<D>`, and given that either would free us of this hack - I would rather work on the latter. :sweat_smile:

Reply:
So what needs to be done to get rid of this double implementation in anvil?

Reply:
I took a look at changing `Xwm`'s hard dependency on calloop, and it was a bit too tangled up for me. I would love some help with that, or with ditching `CalloopData` in anvil. Either would remove the need for this hack.